forecasttools/idata_w_dates_to_df.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+import polars as pl
+
+
+def add_dates_as_coords_to_idata(
+    idata_wo_dates: az.InferenceData,
+    group_dim_dict: dict[str, str],
+    start_date_iso: str,
+) -> az.InferenceData:
+    """
+    Modifies the provided idata object by assigning
+    date arrays to each group specified in group_dim_dict.
+
+    Parameters
+    ----------
+    idata
+        The InferenceData object that contains
+        multiple groups (e.g., observed_data,
+        posterior_predictive).
+    group_dim_dict
+        A dictionary that maps InferenceData group
+        names (e.g., "observed_data", "posterior_predictive"
+        ) to dimension names (e.g., "obs_dim_0").
+    start_date_iso
+        The start date in ISO format (YYYY-MM-DD) from
+        which to begin the date range for the dimension.
+
+    Returns
+    -------
+    idata
+        The modified InferenceData object with
+        date coordinates assigned to each group.
+    """
+
+    # NOTE: failure mode is if groups don't have the same
+    # start date
+
+    # convert received start date string to datetime object
+    start_date_as_dt = datetime.strptime(start_date_iso, "%Y-%m-%d")
+    # copy idata object to avoid modifying the original
+    idata_w_dates = idata_wo_dates.copy()
+    # modify indices of each selected group to dates
+    for (
+        group_name,
+        dim_name,
+    ) in group_dim_dict.items():
+        idata_group = getattr(idata_w_dates, group_name, None)
+        if idata_group is not None:
+            interval_size = idata_group.sizes[dim_name]
+            interval_dates = (
+                pl.DataFrame()
+                .select(
+                    pl.date_range(
+                        start=start_date_as_dt,
+                        end=start_date_as_dt
+                        + pl.duration(days=interval_size - 1),
+                        interval="1d",
+                        closed="both",
+                    )
+                )
+                .to_series()
+                .to_numpy()
+                .astype("datetime64[ns]")
+            )
+            idata_group_with_dates = idata_group.assign_coords(
+                {dim_name: interval_dates}
+            )
+            setattr(
+                idata_w_dates,
+                group_name,
+                idata_group_with_dates,
+            )
+        else:
+            print(f"Warning: Group '{group_name}' not found in idata.")
+    return idata_w_dates
+
+
+def idata_forecast_w_dates_to_df(
+    idata_w_dates: az.InferenceData,
+    location: str,
+    postp_val_name: str,
+    postp_dim_name: str,
+    timepoint_col_name: str = "date",
+    value_col_name: str = "hosp",
+) -> pl.DataFrame:
+    """
+    Converts an Arviz InferenceData object
+    into a Polars dataframe contains dates
+    and draws for a given location. The
+    number of rows in the dataframe is equal
+    to the number of posterior predictive
+    samples times the duration of the time
```

Comment:
What happens to variables that don't change in time?

This is not the typical tidy_draws format, but I'm not sure how the usual format will play with date indexes.

Reply:
It will work fine with date indexes:

``` r
library(tidybayes)
library(tidyverse)
set.seed(5)
n = 10

example <- 
  tibble(
    condition = factor(rep(str_c("A[", seq(today()-4, today(), by = 1), "]"), n)),
    response = rnorm(n * 5, c(0,1,2,1,-1), 0.5)
  ) %>%
  group_by(condition) %>%
  mutate(.chain = 1, .iteration = row_number()) %>% 
  pivot_wider(names_from = condition, values_from = response) %>% 
  mutate(.draw = tidybayes:::draw_from_chain_and_iteration_(.chain, .iteration), .after = .iteration) %>%  tidy_draws()


example %>% spread_draws(A[date])
#> # A tibble: 50 × 5
#> # Groups:   date [5]
#>    date             A .chain .iteration .draw
#>    <chr>        <dbl>  <int>      <int> <int>
#>  1 2024-10-21 -0.420       1          1     1
#>  2 2024-10-21 -0.301       1          2     2
#>  3 2024-10-21  0.614       1          3     3
#>  4 2024-10-21 -0.0695      1          4     4
#>  5 2024-10-21  0.450       1          5     5
#>  6 2024-10-21 -0.147       1          6     6
#>  7 2024-10-21  0.158       1          7     7
#>  8 2024-10-21  0.476       1          8     8
#>  9 2024-10-21  0.775       1          9     9
#> 10 2024-10-21  0.281       1         10    10
#> # ℹ 40 more rows
```

<sup>Created on 2024-10-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>